clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java

```diff
@@ -1896,6 +1902,185 @@ private void recordRegionHybridConsumptionStats(
+          newState = LEADER_COMPLETE_STATE_UNKNOWN;
+        }
+
+        if (oldState != newState) {
+          LOGGER.info(
+              "LeaderCompleteState for store {} version {} partition {} changed from {} to {}",
+              storeName,
+              versionNumber,
+              partitionConsumptionState.getPartition(),
+              oldState,
+              newState);
+          partitionConsumptionState.setLeaderCompleteState(newState);
+        } else {
+          LOGGER.debug(
+              "LeaderCompleteState for store {} version {} partition {} received from leader: {} and is unchanged from the previous state",
+              storeName,
+              versionNumber,
+              partitionConsumptionState.getPartition(),
+              newState);
+        }
+        if (!partitionConsumptionState.isFirstHeartBeatSOSReceived()) {
+          partitionConsumptionState.setFirstHeartBeatSOSReceived(true);
+        }
+      }
+    }
+  }
+
+  /**
+   * Leaders propagate HB SOS message from RT to local VT (to all subpartitions in case if amplification
+   * Factor is configured to be more than 1) with updated LeaderCompleteState header:
+   * Adding the headers during this phase instead of adding it to RT directly simplifies the logic
+   * of how to identify the HB SOS from the correct version or whether the HB SOS is from the local
+   * colo or remote colo, as the header inherited from an incorrect version or remote colos might
+   * provide incorrect information about the support of the header and the leader state.
+   */
+  private void propagateHeartbeatFromUpstreamTopicToLocalVersionTopic(
+      PartitionConsumptionState partitionConsumptionState,
+      PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long> consumerRecord,
+      LeaderProducedRecordContext leaderProducedRecordContext,
+      int partition,
+      String kafkaUrl,
+      int kafkaClusterId,
+      long beforeProcessingRecordTimestampNs) {
+    LeaderProducerCallback callback = createProducerCallback(
+        consumerRecord,
+        partitionConsumptionState,
+        leaderProducedRecordContext,
+        partition,
+        kafkaUrl,
+        beforeProcessingRecordTimestampNs);
+    LeaderMetadataWrapper leaderMetadataWrapper = new LeaderMetadataWrapper(consumerRecord.getOffset(), kafkaClusterId);
+    List<Integer> subPartitions =
+        PartitionUtils.getSubPartitions(partitionConsumptionState.getUserPartition(), amplificationFactor);
+    LeaderCompleteState leaderCompleteState =
+        LeaderCompleteState.getLeaderCompleteState(partitionConsumptionState.isCompletionReported());
+    /**
+     * The maximum value between the original producer timestamp and the timestamp when the message is added to the RT topic is used:
+     * This approach addresses scenarios wrt clock drift where the producer's timestamp is consistently delayed by several minutes,
```

Comment:
Cool! But I am curious; is it defensive coding, or did you really encounter this problem in tests somehow?

Reply:
it is defensive for now, discussed in the design doc.